### PR TITLE
WT-13310 Random Cursor

### PR DESCRIPTION
**Pull request by ershov**
This PR fixes the issue where the session RNG returns the same sequence of numbers.

1. Because the session's `rnd` uses a carefully crafted initial seed to work best with skiplists, this PR introduces another per-session random generator for general purposes. To avoid confusion, these generators are given specific names: `rnd_skiplist` and `rnd_random`.

2. Functions have been renamed for better clarity and readability:
   * `__wt_random_init_seed` is renamed to `__wt_session_rng_init_once`, and its implementation is modified:
       * It will initialize the session's RNG at most once (using `WT_SESSION_FIRST_USE()`).
       * It requires `session->id` to be set before the call to this function.
       * `rnd_skiplist` is initialized the same way as before, maintaining the specialized seed.
       * `rnd_random` is initialized with a uniquely selected seed to ensure it never overlaps within the same application.
   * `__wt_random_init` is renamed to `__wt_random_init_default` to highlight that it uses the default seed.
   * `__wt_random_init_custom_seed` is renamed to `__wt_random_init_seed` for conciseness - since the user-supplied seed is always a "custom" one.

3. Removed unnecessary `volatile` keywords.

4. Other cleanup of no longer used code and macros.

The new test `test/csuite/random_session/test_random_session` attempts to reproduce the bug in two different ways:
1. It opens a session, generates a random number with its RNG and closes the session (in a loop). In this mode, in fact, the issue doesn't show up because the session's RNG is only initialized once and the same session object is getting reused.
2. Rapidly creates a bunch of sessions and generates a random number with their RNGs. This actually reproduces the issue in the `develop` branch but this PR fixes it.

### CODE REVIEW

File: src/support/rand.c [hunk 1]
```diff
@@ -96,53 +95,33 @@ __wt_random_init_custom_seed(WT_RAND_STATE volatile *rnd_state, uint64_t v)
...
-    M_Z(rnd) =
-      (uint32_t)ts.tv_nsec ^ (uint32_t)WT_LEFT_CIRCULAR_SHIFT32(ts.tv_sec, 27) ^ DEFAULT_SEED_Z;
-/*
- * Some system clocks do not have a high enough resolution between each tick cycle. Perform an extra
- * xor against the machine's timestamp counter.
- */
-#ifdef _WIN32
-    rnd.v ^= __wt_rdtsc();
-#endif
-    rnd.v ^= (uint64_t)threadid;
```
Comment (jiechenbo) [inline src/support/rand.c:132]: I believe we did this calculation on purpose, following George Marsaglia's recommendation: https://www.jstatsoft.org/article/view/v008i14/916

File: src/support/rand.c [hunk 2]
```diff
@@ -185,3 +164,19 @@ __wt_random(WT_RAND_STATE volatile *rnd_state) WT_GCC_FUNC_ATTRIBUTE((visibility
...
+/*
+ * __wt_random_init --
+ *     Initialize the state of a 32-bit pseudo-random number by a session's rng.
+ */
+void
+__wt_random_init(WT_SESSION_IMPL *session, WT_RAND_STATE *rnd_state)
+  WT_GCC_FUNC_ATTRIBUTE((visibility("default")))
+{
+    if (session != NULL)
+        __wt_random_init_seed(rnd_state, __wt_random(&session->rnd_random));
```
Comment (jiechenbo) [inline src/support/rand.c:177]: Why do we call both `__wt_random_init_seed` and `__wt_random` here?

**Review by jiechenbo — COMMENTED** [review]
I personally think this is at least a medium risk. Although we have a reproducer for the issue, we are not aware of any other potential risk this change may incur. Furthermore since we changed the skiplist rng number it can incr performance penalties in case we accidentally forgot a piece of code that was related to RNG and skiplists.

**Comment by ershov** [comment]
> I personally think this is at least a medium risk. Although we have a reproducer for the issue, we are not aware of any other potential risk this change may incur. Furthermore since we changed the skiplist rng number it can incr performance penalties in case we accidentally forgot a piece of code that was related to RNG and skiplists.

In fact, this PR keeps the skiplist’s RNG intact, while introducing a separate RNG for other purposes. What kind of risks could we expect?

File: src/support/rand.c (hunk 1 above)
Comment (ershov) [reply src/support/rand.c:132]: We now use time and session ID instead - should that be a good replacement for time and thread ID? Do you see any potential issues or edge cases we should watch out for?

File: test/csuite/random_session/main.c [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    /* Initialize database. */
+
+    testutil_work_dir_from_path(home, sizeof(home), "WT_TEST.random_session");
+    testutil_recreate_dir(home);
+
+    testutil_check(wiredtiger_open(home, NULL, conn_config, &conn));
+
+    /* Test one session. */
+
+    for (int cycle = 0; (++cycle) <= N_RETRIES;) {
```
Comment (pmacko86) [inline test/csuite/random_session/main.c:75]: Is there a reason for using this `for` construct, instead of the more common format?
```
for (int cycle = 1; cycle <= N_RETRIES; ++cycle)
```
It makes it harder to read, especially given that it is easy to forget that the condition in the middle gets evaluated also before the first iteration. If possible, the condition should not have side effects.

File: test/csuite/random_session/main.c [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        for (int i = 0; i < N_SESSIONS; i++) {
+            WT_SESSION *session;
+            testutil_check(conn->open_session(conn, NULL, NULL, &session));
+            number = __wt_random(&((WT_SESSION_IMPL *)session)->rnd_random);
+            testutil_check(session->close(session, NULL));
+            if (cycle > 1 && prev_number == number) {
+                if (cycle < N_RETRIES)
+                    /* To eliminate flakiness, give it another go. */
+                    goto retry_single;
+                fprintf(stderr, "Random numbers repeated for %d cycles in single session\n", cycle);
```
Comment (pmacko86) [inline test/csuite/random_session/main.c:88]: I don't understand this test. It creates 10 sessions per cycle, but the comment is about a single session?

If the cycles are meant as retries, I think you're missing `break` like you have in the multiple tests below. IN that case, it may also help to rename "cycle" to something more like "retry" for clarity. The word "cycles" implies that things are repeated for reasons other than retries.

File: test/csuite/random_session/main.c [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                    /* To eliminate flakiness, give it another go. */
+                    goto retry_multi;
+                fprintf(
+                  stderr, "Random numbers repeated for %d cycles in subsequent sessions\n", cycle);
+                random_numbers_repeated = true;
+            }
+            prev_number = number;
+        }
+
+        /* Check is any random numbers repeat. */
```
Comment (pmacko86) [inline test/csuite/random_session/main.c:123]: "is" -> "if"

File: bench/tiered/push_pull.c [hunk 6]
```diff
@@ -315,13 +315,12 @@ recover_validate(const char *home, uint32_t num_records, uint64_t file_size, uin
     testutil_wiredtiger_open(opts, home, buf, NULL, &conn, true, true);
     testutil_check(conn->open_session(conn, NULL, NULL, &session));
 
-    /* Seed the random number generator */
-    v = (uint32_t)getpid() + num_records + (2 * counter);
-    __wt_random_init_custom_seed(&rnd, v);
```
Comment (pmacko86) [inline bench/tiered/push_pull.c:320]: This change does not seem correct. I haven't spent that much time looking at this test, but it seems to me that the original test intentionally had the seed initialized in the same way in both `populate` and `recover_validate`. cc: @ddanderson

File: src/support/rand.c (hunk 1 above)
Comment (pmacko86) [reply src/support/rand.c:132]: Don't know this from top of my head—do we reuse session IDs? If so, would that cause any problems?

File: src/support/rand.c [hunk 7]
```diff
@@ -96,53 +95,33 @@ __wt_random_init_custom_seed(WT_RAND_STATE volatile *rnd_state, uint64_t v)
...
-#endif
-    rnd.v ^= (uint64_t)threadid;
-    rnd.v ^= rnd.v << 13;
-    rnd.v ^= rnd.v >> 7;
-    rnd.v ^= rnd.v << 17;
+    __wt_random_init_default(&session->rnd_skiplist);
 
-    *rnd_state = rnd;
+    __wt_random_init_seed(&session->rnd_random,
+      ((uint64_t)session->id + 1) * __wt_clock(session) / WT_BILLION + (uint64_t)getpid());
```
Comment (pmacko86) [inline src/support/rand.c:116]: Is there a reason why we don't include `__wt_clock(session) % WT_BILLION` as you had `(t % WT_BILLION)` below? I like the idea of capturing the low bits of `rdtsc` (which I think we use as the default implementation for `__wt_clock`?).

**Review by pmacko86 — COMMENTED** [review]
Thank you for putting this big PR together. I like the new names for the various random functions, and I also like the idea of splitting the random number generator state between skiplists and everything else.

I left a few questions for you in the comments, and I highlighted a few potential issues for you to evaluate. But overall, this is on the right track.

File: bench/tiered/push_pull.c (hunk 6 above)
Comment (ddanderson) [reply bench/tiered/push_pull.c:320]: Right, nice catch.  I think they require the same seed, as "random" data is stored, and later retrieved and checked.  Since push_pull is a benchmark, it's probably never run by PR tests.  I don't know how easy it is to run to ensure that nothing is broken.  cc: @raviprakashgiri29

File: src/support/rand.c (hunk 1 above)
Comment (ershov) [reply src/support/rand.c:132]: No, since the RNG is only initialized once in the lifetime of the application for a given session id (which equals to the session's index in the session array).

File: src/support/rand.c [hunk 8]
```diff
@@ -96,53 +95,33 @@ __wt_random_init_custom_seed(WT_RAND_STATE volatile *rnd_state, uint64_t v)
...
 }
 
 /*
  * __wt_random --
  *     Return a 32-bit pseudo-random number.
  */
 uint32_t
-__wt_random(WT_RAND_STATE volatile *rnd_state) WT_GCC_FUNC_ATTRIBUTE((visibility("default")))
+__wt_random(WT_RAND_STATE *rnd_state) WT_GCC_FUNC_ATTRIBUTE((visibility("default")))
```
[the remark this answers is not in the document]
Comment (ershov) [reply src/support/rand.c:124]: I believe, these are leftovers from incorrect attempts to handle multithreading. In the modern world we do it via atomics and barriers.

File: test/csuite/random_session/main.c (hunk 5 above)
Comment (ershov) [reply test/csuite/random_session/main.c:123]: Thanks!

File: bench/tiered/push_pull.c (hunk 6 above)
Comment (ershov) [reply bench/tiered/push_pull.c:320]: Good catch! Reverted this.

File: test/csuite/random_session/main.c (hunk 4 above)
Comment (ershov) [reply test/csuite/random_session/main.c:88]: Yes, there's a missing `break`, thanks for spotting!
Also, renamed `cycles` to `retries` for clarity.

Comment (ershov) [reply test/csuite/random_session/main.c:88]: The first test generates a random number in N sessions sequentially, with only one session open at a time.
The second test generates random numbers with N sessions open simultaneously.

File: src/support/rand.c [hunk 9]
```diff
@@ -96,53 +95,34 @@ __wt_random_init_custom_seed(WT_RAND_STATE volatile *rnd_state, uint64_t v)
...
+        return;
 
     /*
-     * Use this, instead of __wt_random_init, to vary the initial state of the RNG. This is
-     * (currently) only used by test programs, where, for example, an initial set of test data is
-     * created by a single thread, and we want more variability in the initial state of the RNG.
-     *
-     * Take the seconds and nanoseconds from the clock together with the thread ID to generate a
-     * 64-bit seed, then smear that value using algorithm "xor" from Marsaglia, "Xorshift RNGs".
+     * Session's skiplist RNG is initialized with the special default seed.
```
Comment (ddanderson) [inline src/support/rand.c:111]: Hmm, that means that every session has an RNG that is used exclusively for skip lists, and all sessions have the same starting seed.  That doesn't seem right to me, we could get a lot of synchronous behavior on the skiplist RNGs for a while.  If there's a reason to insulate the two RNGs from another, that's okay, but I think we should have different starting points for all sessions.

**Review by ddanderson — COMMENTED** [review]
Overall, lots of the details look good to me.
I'm just worried about the skiplist RNG having synchronized behavior.

File: src/support/rand.c (hunk 9 above)
Comment (ershov) [reply src/support/rand.c:111]: I also don't think this is good but since I was requested to preserve this RNG for skiplists, I did so.
I do believe skiplists should use the normal RNG, that can be done as a separate work.

Comment (pmacko86) [reply src/support/rand.c:111]: This is a good question for @agorrod.

The problem with Marsaglia's random number generator is that if you start with a bad initial seed, you would ultimately end up with degenerate skip lists, which would hurt performance. It would happen in a small minority of cases, but given the scale of our deployment, the probability of _some_ customer getting degenerate skip lists _at some point_ is non-trivial and should be avoided. The starting seed that we use is known to _always_ result in generating good randomness for our skip lists.

On the other hand, I don't think every session starting with the same generator is a problem: That's what we currently do, and the non-skiplist use cases are not that many, so if there was a problem with always having the same starting seed, we would have seen this by now.

File: test/csuite/random_session/main.c (hunk 4 above)
Comment (pmacko86) [reply test/csuite/random_session/main.c]: This makes sense—could you please add these explanations as comments?

File: test/csuite/random_session/main.c (hunk 3 above)
Comment (ershov) [reply test/csuite/random_session/main.c]: I've fixed this, thanks for spotting!

**Review by pmacko86 — APPROVED** [review]
LGTM. This addresses all my comments. Thank you!

File: src/support/rand.c (hunk 9 above)
Comment (ddanderson) [reply src/support/rand.c:115]: @ershov pointed out that this is the way it currently works for skiplists.  I have some concerns, but they don't have to do with this PR, which successfully fixes an existing problem.

**Review by ddanderson — APPROVED** [review]
LGTM

**Review by jiechenbo — APPROVED** [review]
LGTM! Thanks for making the change
